### ISSUE TITLE
FileNotFoundError in `dask.array.to_zarr`, when updating from zarr=2.13.3 to zarr=2.13.6.

### ISSUE THREAD

**Issue by tcompa** [dask/dask#9841]
**Describe the issue**:

Hi there, I found what looks like a bug using `dask.array.to_zarr` with zarr=2.13.6. If you think the issue belongs to https://github.com/zarr-developers/zarr-python, feel free to close this one and I will re-open one over there.

**Minimal Complete Verifiable Example**:
The following code works with dask=2022.7 and zarr=2.13.3, but fails with dask=2022.7 and zarr=2.13.6:
```python
import zarr
import dask
import dask.array as da

print(f"{zarr.__version__=}")
print(f"{dask.__version__=}")

x = da.ones((1, 1))
x.to_zarr("/tmp/ones.zarr", dimension_separator="/")
print("Success")
```

Successful run:
```
zarr.__version__='2.13.3'
dask.__version__='2022.7.0'
Success
```

Failed run
```python traceback
zarr.__version__='2.13.6'
dask.__version__='2022.7.0'
Traceback (most recent call last):
  File "/tmp/mwe.py", line 9, in <module>
    x.to_zarr("/tmp/ones.zarr", dimension_separator="/")
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/dask/array/core.py", line 2828, in to_zarr
    return to_zarr(self, *args, **kwargs)
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/dask/array/core.py", line 3591, in to_zarr
    return arr.store(z, lock=False, compute=compute, return_stored=return_stored)
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/dask/array/core.py", line 1752, in store
    r = store([self], [target], **kwargs)
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/dask/array/core.py", line 1226, in store
    compute_as_if_collection(Array, store_dsk, map_keys, **kwargs)
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/dask/base.py", line 347, in compute_as_if_collection
    return schedule(dsk2, keys, **kwargs)
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/dask/threaded.py", line 89, in get
    results = get_async(
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/dask/local.py", line 511, in get_async
    raise_exception(exc, tb)
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/dask/local.py", line 319, in reraise
    raise exc
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/dask/local.py", line 224, in execute_task
    result = _execute_task(task, data)
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/dask/core.py", line 119, in _execute_task
    return func(*(_execute_task(a, cache) for a in args))
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/dask/array/core.py", line 4260, in store_chunk
    return load_store_chunk(x, out, index, lock, return_stored, False)
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/dask/array/core.py", line 4242, in load_store_chunk
    out[index] = x
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/zarr/core.py", line 1373, in __setitem__
    self.set_basic_selection(pure_selection, value, fields=fields)
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/zarr/core.py", line 1468, in set_basic_selection
    return self._set_basic_selection_nd(selection, value, fields=fields)
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/zarr/core.py", line 1772, in _set_basic_selection_nd
    self._set_selection(indexer, value, fields=fields)
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/zarr/core.py", line 1844, in _set_selection
    self._chunk_setitems(lchunk_coords, lchunk_selection, chunk_values,
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/zarr/core.py", line 2046, in _chunk_setitems
    self.chunk_store.setitems(to_store)
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/zarr/storage.py", line 1394, in setitems
    self.map.setitems(values)
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/fsspec/mapping.py", line 116, in setitems
    self.fs.pipe(values)
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/fsspec/spec.py", line 757, in pipe
    self.pipe_file(self._strip_protocol(k), v, **kwargs)
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/fsspec/spec.py", line 736, in pipe_file
    with self.open(path, "wb", **kwargs) as f:
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/fsspec/spec.py", line 1106, in open
    f = self._open(
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/fsspec/implementations/local.py", line 175, in _open
    return LocalFileOpener(path, mode, fs=self, **kwargs)
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/fsspec/implementations/local.py", line 273, in __init__
    self._open()
  File "/home/tommaso/miniconda3/envs/test/lib/python3.9/site-packages/fsspec/implementations/local.py", line 278, in _open
    self.f = open(self.path, mode=self.mode)
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/ones.zarr/0/0'
```

**Anything else we need to know?**:

* The script works if I replace `dimension_separator="/"` with `dimension_separator="."`.
* [This one](https://github.com/fractal-analytics-platform/fractal-tasks-core/issues/280) is the related issue in our repository (not much more info than in the minimal-failing-example above).
* My first attempts to pin down the source of the error lead me to https://github.com/zarr-developers/zarr-python/pull/1304, but I could not move further (and it may very well not be the actual error origin).

**Environment**:

- Dask version: 2022.7 (I don't think it's crucial though)
- Python version: 3.9.0
- Operating System: Ubuntu 22.04
- Install method (conda, pip, source): conda (in a fresh environment, where I only added dask and zarr)

**Comment by tcompa** [comment]
Addendum: it looks like the issue was solved at least with dask 2023.1:
```
zarr.__version__='2.13.6'
dask.__version__='2023.1.0'
Success
```
Apologies for not checking earlier!

Feel free to keep open or close, depending on what's best for dask.

**Comment by jrbourbeau** [comment]
Thanks for reporting @tcompa! 

As you mentioned, the latest `dask=2023.1.0` release should fix this. There's also https://github.com/zarr-developers/zarr-python/pull/1309 which should help fix things for older `dask` versions.
